File: omni-runtime/stdlib/julia/omni_bridge.jl
# ============================================================
# OMNI Julia Bridge — HPC/SIMD Compute Nexus
# ============================================================
# Layer: Compute (SIMD, Numerical, HPC)
# Julia menggunakan ccall bawaan — TIDAK perlu Pkg.add apapun.
# ============================================================

module OmniBridge

# Deteksi platform
const LIB_EXT = Sys.iswindows() ? "omni_core.dll" :
                Sys.isapple()   ? "libomni_core.dylib" :
                                  "libomni_core.so"

const LIB_PATH = joinpath(@__DIR__, "..", "..", "core", "target", "release", LIB_EXT)

"""
    execute(function_name::String, data::Vector{UInt8}) -> Ptr{Cvoid}

Memanggil fungsi OMNI melalui batas C-ABI dengan zero-copy.
Julia ccall menjamin NO GC PAUSE selama cross-language call.
"""
function execute(function_name::String, data::Vector{UInt8})::Ptr{Cvoid}
    if isempty(data)
        error("[OMNI-E601] Data tidak boleh kosong")
    end

    # ccall langsung ke pointer — tanpa serialisasi
    result_ptr = ccall(
        (:__omni_ffi, LIB_PATH),
        Ptr{Cvoid},                      # return type
        (Ptr{UInt8}, Csize_t),           # arg types
        pointer(data), length(data)       # args
    )

    if result_ptr == C_NULL
        error("[OMNI-E602] Kernel mengembalikan null pointer")
    end

    return result_ptr
end

"""
    execute_simd(data::Vector{Float64}) -> Ptr{Cvoid}

Optimized path: kirim array numerik langsung ke Rust SIMD engine.
Julia Vector{Float64} layout identik dengan Rust Vec<f64> — true zero-copy.
"""
function execute_simd(data::Vector{Float64})::Ptr{Cvoid}
    raw_bytes = reinterpret(UInt8, data)
    return execute("julia_simd_dispatch", raw_bytes)
end

export execute, execute_simd

end # module OmniBridge

File: omni-runtime/stdlib/ruby/omni_bridge.rb
# ============================================================
# OMNI Ruby Bridge — Domain DSL Nexus
# ============================================================
# Layer: Domain (Convention-over-Config, Routing, DSL)
# Menggunakan Ruby Fiddle bawaan — TIDAK perlu `gem install ffi`.
# ============================================================

require 'fiddle'
require 'fiddle/import'

module OmniBridge
  extend Fiddle::Importer

  # Deteksi platform
  LIB_EXT = case RUBY_PLATFORM
            when /mswin|mingw/ then 'omni_core.dll'
            when /darwin/      then 'libomni_core.dylib'
            else                    'libomni_core.so'
            end

  LIB_PATH = File.join(__dir__, '..', '..', 'core', 'target', 'release', LIB_EXT)

  unless File.exist?(LIB_PATH)
    raise LoadError, "[OMNI-E401] Kernel tidak ditemukan: #{LIB_PATH}\nKompilasi: omni build --release"
  end

  dlload LIB_PATH

  # C-ABI function signature
  extern 'void* __omni_ffi(void*, size_t)'

  # Eksekusi fungsi OMNI dari Ruby DSL
  def self.execute(function_name, data_string)
    ptr = Fiddle::Pointer[data_string]
    result_ptr = __omni_ffi(ptr, data_string.bytesize)

    if result_ptr.null?
      raise RuntimeError, "[OMNI-E402] Kernel mengembalikan null pointer"
    end

    result_ptr
  end

  # DSL routing shorthand untuk rb::route
  def self.route(path, &block)
    # Didelegasikan ke OMNI HTTP Router (Phase 12)
    execute("rb_route_register", "#{path}")
    block.call if block_given?
  end
end
